src/callbacks_stage/a_posteriori_limiter.jl

```diff
@@ -0,0 +1,53 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+"""
+    APosterioriLimiter()
+
+Perform antidiffusive stage for a posteriori IDP limiting.
+
+!!! warning "Experimental implementation"
+    This is an experimental feature and may change in future releases.
+"""
+struct APosterioriLimiter end
```

Comment:
- Is there a reference for this?
- Is there a slightly less general name clarifying what it does? For example, the Zhang-Shu limiter is also an a posterior limiter...
- Is it required to make this experimental right now?
  - Do you have a roadmap to declare it not experimental anymore?
  - What else do you plan to do with this?

Reply:
> Is there a reference for this?

Since it is only required for `IndicatorIDP`, I will mention this here correctly. In addition, yes, it's probably useful to add the Paper references here too.

> Is there a slightly less general name clarifying what it does? For example, the Zhang-Shu limiter is also an a posterior limiter...

We briefly discussed it in [here](https://github.com/trixi-framework/Trixi.jl/pull/1476#discussion_r1215349172). “Limiter” seemed to be more helpful than “antidiffusive_stage” and then we only distinguished between IDP and MCL. But you're right. Maybe it's a good idea to add IDP somewhere. `APosterioriIDPLimiter`?

> Is it required to make this experimental right now? 

I think I will not have to change anything in here in the future. But I'm fine with keeping the warning until all IDP limiting features are merged to main. Possibly we still find some issues.

Reply:
Okay, it's fine to keep it experimental right now.

Is there some name we can give to this kind of limiters - something like `PositivityPreservingLimiterZhangShu`? Could it be `InvariantDomainPreservingLimiterWhaterver`?

Reply:
Yeah, "limiter" isn't very specific - you can have Zhang-Shu, minmod, subcell a-posteriori...

What about "blending" or "blending limiter" in the function name?

Reply:
Since I updated the docstring with additional information and references to
```
    APosterioriLimiter()

Perform antidiffusive stage for the a posteriori IDP limiter called with
[`VolumeIntegralSubcellLimiting`](@ref) using [`IndicatorIDP`](@ref).

!!! warning "Experimental implementation"
    This is an experimental feature and may change in future releases.

## References

- Rueda-Ramírez, Pazner, Gassner (2022)
  Subcell Limiting Strategies for Discontinuous Galerkin Spectral Element Methods
  [DOI: 10.1016/j.compfluid.2022.105627](https://doi.org/10.1016/j.compfluid.2022.105627)
- Pazner (2020)
  Sparse invariant domain preserving discontinuous Galerkin methods with subcell convex limiting
  [DOI: 10.1016/j.cma.2021.113876](https://doi.org/10.1016/j.cma.2021.113876)
```
I think something like `APosterioriIDPLimiter` or `APosterioriAntidiffusive(IDP)Limiter` should be a good name.

In this context, we can also think about renaming the “indicators” (`IndicatorIDP` and `IndicatorMCL`) to “limiters”. Then, it differs from the naming of “IndicatorHennemannGassner”, but this seems to make sense to me in this context. What do you think? @ranocha @amrueda @sloede @jlchan 

Then, maybe use the callback `APosterioriIDPLimiterAntidiffusiveStage` for the limiter `APosterioriIDPLimiter`.

Reply:
> `APosterioriIDPLimiterAntidiffusiveStage` is kind of long

I thought so too😅

To summarize the current idea.
We got the volume integral `VolumeIntegralSubcellLimiting`.
It uses the limiter `APosterioriIDPLimiter` (instead of `IndicatorIDP`) and (in future PRs) `MonolithicConvexLimiter` (instead of `IndicatorMCL`).
The limiter `APosterioriIDPLimiter` requires an antidiffusive stage, which is implemented as a stage callback named either:
- `APosterioriIDPLimiterAntidiffusive`,
- `APosterioriIDPLimiterAntidiffusion`,
- `APosterioriIDPLimiterBlending`.

I think we are good with those names.

Reply:
After renaming the indicators to limiters, it is not really useful to define them as `AbstractIndicator`s anymore. Therefore, I would add the new type `AbstractLimiter`.
Furthermore, this provides the possibility to move all limiter specific routines to new files, e.g., `limiter.jl` and `limiter_2d.jl`. This way, `dg_2d` and `indicators_2d.jl` will not become too large as it happens now:
- `dgsem_tree/dg_2d` from 1080 to 2381 lines and `dgsem_structured/dg_2d` from 564 to 841 lines.
- `dgsem_tree/indicators_2d` from 546 to 1206 lines and `dgsem_structured/indicators_2d` from 30 to 232 lines

Do you agree with these changes? @ranocha @sloede

Reply:
Do the new "limiters" use the same interface as the "indicators" we currently have in Trixi.jl? If not, then this is a great change.

Reply:
What do you mean by “same interface”?
Do you mean how they are called in the first place? If so, it is very similar rn compared to the shock capturing function of `IndicatorHennemannGassner`), since both are parameters of the specific volume integral.

But since `IndicatorHennemannGassner` and `VolumeIntegralShockCapturingHG` only work in this combination, the name “indicator” fits well here. For the subcell limiting, we have the `VolumeIntegralSubcellLimiting`, where “indicator” doesn't really work because it is more than just an indicator (and we have more than 1 option). That's why I change it to “limiter” here as well, such that we can call `volume_integral.limiter` and can initialize like follows:
```
limiter = APosterioriIDPLimiter(equations, basis;
                                positivity=true, variables_cons=(Trixi.density,))

volume_integral = VolumeIntegralSubcellLimiting(limiter;
                                                volume_flux_dg=volume_flux,
                                                volume_flux_fv=surface_flux)
```

Reply:
We have several indicators we use in Trixi.jl, e.g., to control the shock capturing in `VolumeIntegralShockCapturingHG` or adaptive mesh refinement. If your new `limiter = APosterioriIDPLimiter` satisfies the same interface - same kind of input and output - I think it will be confusing to call it `limiter` (like the positivity preserving limiter we have) and not `indicator`.

Reply:
Could you please put it on the agenda of the next Trixi.jl meeting? It's an interface/usability decision more people should be involved in